### PR TITLE
Laboratory work #3, Timofey Baranov - 22FPL1

### PR DESCRIPTION
**Pull request by tima-baranov**

### CODE REVIEW

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Nothing to review

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Nothing to review(

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Nothing to review

File: lab_3_generate_by_ngrams/main.py [hunk 1]
```diff
@@ -159,6 +226,17 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
 
         In case of corrupt input arguments, None is returned
         """
+        if not isinstance(decoded_corpus, tuple) or not decoded_corpus:
+            return None
+        postprocessed_text = ''
+        for index, token in enumerate(decoded_corpus):
+            if token == self._end_of_word_token and index == len(decoded_corpus) - 1:
+                postprocessed_text += '.'
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:234]: fstring

File: lab_3_generate_by_ngrams/main.py [hunk 2]
```diff
@@ -159,6 +226,17 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
...
+            elif token == self._end_of_word_token:
+                postprocessed_text += ' '
+            else:
+                postprocessed_text += token
```
Comment (artyomtugaryov) [inline lab_3_generate_by_ngrams/main.py:238]: fstring

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_generate_by_ngrams/main.py [hunk 3]
```diff
@@ -41,6 +43,19 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
+            return None
+        tokenized_text = []
+        for token in text.lower():
+            if token.isalpha():
+                tokenized_text.append(token)
+            elif token.isspace() and tokenized_text[-1] != self._end_of_word_token:
+                tokenized_text.append(self._end_of_word_token)
+        if not text[-1].isalnum() and tokenized_text[-1] != self._end_of_word_token:
+            tokenized_text.append(self._end_of_word_token)
+        if len(tokenized_text) == tokenized_text.count(self._end_of_word_token):
```
[the remark this answers is not in the document]
Comment (tima-baranov) [reply lab_3_generate_by_ngrams/main.py:56]: fixed

File: lab_3_generate_by_ngrams/main.py [hunk 4]
```diff
@@ -76,6 +95,12 @@ def get_token(self, element_id: int) -> Optional[str]:
 
         In case of corrupt input arguments or arguments not included in storage, None is returned
         """
+        if not isinstance(element_id, int) or not element_id in self._storage.values():
+            return None
```
[the remark this answers is not in the document]
Comment (tima-baranov) [reply lab_3_generate_by_ngrams/main.py]: fixed

File: lab_3_generate_by_ngrams/main.py [hunk 5]
```diff
@@ -93,6 +118,19 @@ def encode(self, text: str) -> Optional[tuple[int, ...]]:
         In case of corrupt input arguments, None is returned.
         In case any of methods used return None, None is returned.
         """
+        if not isinstance(text, str) or not text:
+            return None
+        tokenized_text = self._tokenize(text)
+        encoded_text = []
+        if tokenized_text is None:
+            return None
+        for token in tokenized_text:
```
[the remark this answers is not in the document]
Comment (tima-baranov) [reply lab_3_generate_by_ngrams/main.py:127]: fixed

File: lab_3_generate_by_ngrams/main.py [hunk 6]
```diff
@@ -104,6 +142,8 @@ def _put(self, element: str) -> None:
         In case of corrupt input arguments or invalid argument length,
         an element is not added to storage
         """
+        if isinstance(element, str) and len(element) == 1 and element not in self._storage:
```
[the remark this answers is not in the document]
Comment (tima-baranov) [reply lab_3_generate_by_ngrams/main.py]: fixed

File: lab_3_generate_by_ngrams/main.py (hunk 1 above)
Comment (tima-baranov) [reply lab_3_generate_by_ngrams/main.py]: fixed

File: lab_3_generate_by_ngrams/main.py [hunk 7]
```diff
@@ -159,6 +226,17 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
 
         In case of corrupt input arguments, None is returned
         """
+        if not isinstance(decoded_corpus, tuple) or not decoded_corpus:
+            return None
+        postprocessed_text = ''
+        for index, token in enumerate(decoded_corpus):
+            if token == self._end_of_word_token and index == len(decoded_corpus) - 1:
+                postprocessed_text += '.'
+            elif token == self._end_of_word_token:
```
[the remark this answers is not in the document]
Comment (tima-baranov) [reply lab_3_generate_by_ngrams/main.py]: fixed

File: lab_3_generate_by_ngrams/main.py [hunk 8]
```diff
@@ -159,6 +226,17 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> Optional
...
+                postprocessed_text += ' '
```
[the remark this answers is not in the document]
Comment (tima-baranov) [reply lab_3_generate_by_ngrams/main.py]: fixed

File: lab_3_generate_by_ngrams/main.py (hunk 2 above)
Comment (tima-baranov) [reply lab_3_generate_by_ngrams/main.py]: fixed

File: lab_3_generate_by_ngrams/main.py [hunk 9]
```diff
@@ -41,6 +43,19 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
         """
+        if not isinstance(text, str):
+            return None
+        tokenized_text = []
+        for token in text.lower():
+            if token.isalpha():
+                tokenized_text.append(token)
+            elif token.isspace() and tokenized_text[-1] != self._end_of_word_token:
+                tokenized_text.append(self._end_of_word_token)
+        if not text[-1].isalnum() and tokenized_text[-1] != self._end_of_word_token:
```
[the remark this answers is not in the document]
Comment (tima-baranov) [reply lab_3_generate_by_ngrams/main.py:54]: fixed

File: lab_3_generate_by_ngrams/main.py [hunk 10]
```diff
@@ -129,6 +178,13 @@ def fill_from_ngrams(self, content: dict) -> None:
         Args:
             content (dict): ngrams from external JSON
         """
+        if not isinstance(content, dict) or not content:
+            return None
+        for key in content['freq']:
+            for element in key:
+                if element.isalpha():
```
[the remark this answers is not in the document]
Comment (tima-baranov) [reply lab_3_generate_by_ngrams/main.py:184]: fixed

**Pull request closed without merging**